Add LSP support for showing Macro Expansions

lokesh-tr opened the pull request:
**Note: This Feature is Experimental 🧪. You have to first enable this feature by passing `--experimental-feature show-macro-expansions` to `sourcekit-lsp`**

This PR aims to bring a streamlined way of showing macro expansions through the LSP.

This allows users to pick a code action on a macro which performs an `ExecuteCommandRequest` with an `ExpandMacroCommand`. The server then generates the macro expansion (using semantic refactoring request) and stores it in a temporary directory. The server also performs a `ShowDocumentRequest` ([LSP specification](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#window_showDocument)) on the client. Thus, displaying the expansion to the users.

`ShowDocumentRequest` being a part of the LSP spec will allow us to make macro expansions accessible across wide range of editors.

The result of this PR is that macro expansions work identical to that of Xcode with the only difference being Xcode showing the expansions inline, while we don't have a standardised LSP spec for displaying inline. (Although some customisation is possible for *VS Code* alone through their API).

#### How it works?
1. `sourcekitd` returns an "Inline Macro" Code Action when the editor requests for the available code actions on a macro.
2. We inject an "Expand Macro" Code Action having a `ExpandMacroCommand` before passing the response of `sourcekitd` to the editor.
3. If the user decides to pick the "Expand Macro" Code Action, then the editor sends an `ExecuteCommandRequest` containing the `ExpandMacroCommand`.
4. Upon receiving the `ExecuteCommandRequest` containing the `ExpandMacroCommand`, We then ask `sourcekitd` for the expansion of the given macro by making a semantic refactoring request.
5. The response is interpreted as a `MacroExpansion`, and for each edit (expansion / replacement) it suggested, we make a temporary file as follows:
`<NSTemporaryDirectory>/sourcekit-lsp/GeneratedMacroExpansions/<edit.bufferName>/<SourceFileName>_<LineColumnRange>.<SourceFileExtension>`

Here,
- `<edit.bufferName>` will be the buffer name associated with each edit, it looks like `@__swiftmacro_..._.swift` without the ".swift" extension at last.
- `<LineColumnRange>` will look like `LaCb-LcCd` where a & c represent Line number of start and end of expansion and b & d represent Column number of start and end of expansion.
6. Once, the temporary file has been created, we make a `ShowDocumentRequest`, containing the URI of temporary file, to the editor.
7. Now, the editor shows the temporary file containing the macro expansion to the user. 🎉

#### Follow-up Work & Future Directions:
1. The current implementation lays foundation for macro expansions of any kind, but `@attached` macro expansions is little buggy. LSP support for `@attached` macro expansions will be fixed in the following PR: https://github.com/apple/sourcekit-lsp/pull/1479
2. This implementation is **not** 1:1 with Xcode's "Expand Macro" feature. The only difference being the fact that we open separate documents for each expansion edit, whereas Xcode shows them all inline. This is a known limitation due to the LSP specifications having no support for showing anything inline the document.
**But for people using VS Code, we are looking forward to implement some custom functionality to show expansions inline with the help of VS Code API.** ~This will be implemented in a future PR.~
(Implemented in 
https://github.com/swiftlang/sourcekit-lsp/pull/1479
https://github.com/swiftlang/vscode-swift/pull/945)

#### References:
 - https://github.com/apple/sourcekit-lsp/issues/755 
 - https://github.com/apple/sourcekit-lsp/pull/892
 - https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#window_showDocument
 - https://docs.swift.org/swift-book/documentation/the-swift-programming-language/macros/
 - https://github.com/apple/swift-evolution/blob/main/proposals/0415-function-body-macros.md
 - https://github.com/apple/swift/pull/66295
 
 Fixes https://github.com/apple/sourcekit-lsp/issues/755
 rdar://110516432
 
[Expansion of Swift Macros in Visual Studio Code - Google Summer Of Code 2024](https://summerofcode.withgoogle.com/programs/2024/projects/zQNf7ztP)
@lokesh-tr @ahoppen @adam-fowler

ahoppen commented on Sources/SourceKitLSP/Swift/Refactoring.swift:
```diff
@@ -0,0 +1,124 @@
+  /// - Parameters:
+  ///   - title: The title of the refactoring action.
+  ///   - dict: Response dictionary to extract information from.
+  ///   - url: The client URL that triggered the `semantic_refactoring` request.
```
I think I fixed this comment mismatch last week 😉 

```suggestion
  ///   - snapshot: The snapshot that triggered the `semantic_refactoring` request.
```

ahoppen commented on Sources/SourceKitLSP/Swift/ExpandMacroCommand.swift:
```diff
@@ -0,0 +1,74 @@
+    )
+  }
+
+  public init(title: String, actionString: String, positionRange: Range<Position>, textDocument: TextDocumentIdentifier)
+  {
```
Nitpick, I prefer to break this on the parameters.

```suggestion
  public init(
    title: String, 
    actionString: String, 
    positionRange: Range<Position>, 
    textDocument: TextDocumentIdentifier
  ) {
```

ahoppen commented on Sources/SourceKitLSP/Swift/Refactoring.swift:
```diff
@@ -0,0 +1,124 @@
+import LanguageServerProtocol
+import SourceKitD
+
+protocol Refactoring {
```
Would `RefactoringResult` or `RefactoringResponse` make more sense as the name here? `Refactoring` sounds more like what you implemented as `RefactorCommand` to me.

ahoppen commented on Sources/SourceKitLSP/Swift/Refactoring.swift:
```diff
@@ -0,0 +1,124 @@
+  ///   - responseType: The response type `T.Type` of the particular command
+  /// - Returns:
+  ///   - The `Refactoring` as `T`
+  func refactoring<T: Refactoring>(_ refactorCommand: any RefactorCommand, _ responseType: T.Type) async throws -> T {
```
Instead of explicitly passing in the response type here, what do you think about adding a 

```swift
associatedtype Result: Refactoring
```

to `RefactorCommand` and then changing the function definition to


```suggestion
  func refactoring<T: RefactoringCommand>(_ refactorCommand: T, _ responseType: T.Result) async throws -> T.Result {
```

ahoppen commented on Sources/SourceKitLSP/Swift/SemanticRefactoring.swift:
```diff
@@ -1,3 +1,4 @@
+import LSPLogging
```
Could you move the import after the copyright header?

ahoppen commented on Sources/sourcekit-lsp/SourceKitLSP.swift:
```diff
@@ -193,6 +193,11 @@ struct SourceKitLSP: AsyncParsableCommand {
   )
   var generatedInterfacesPath = defaultDirectoryForGeneratedInterfaces
 
+  @Option(
+    help: "Specify the directory where generated macro expansions will be stored"
+  )
+  var generatedMacroExpansionsPath = defaultDirectoryForGeneratedMacroExpansions
+
```
Instead of adding options for every new kind of generated file directories, what do you think about generalizing `generatedInterfacesPath` to `generatedFilesPath` and then placing the generated interfaces and the macro expansions in those subdirectories?

ahoppen commented on Sources/SourceKitLSP/Swift/SwiftCommand.swift:
```diff
@@ -15,15 +15,20 @@ import LanguageServerProtocol
   var title: String { get set }
 }
 
+public typealias SwiftCommandHandler = (any SwiftCommand) async throws -> LSPAny
+
```
I think we only use this once so I’d just use the function type inline. Just means that you don’t need to look up what `SwiftCommandHandler`. Just means that you don’t need to look up what `SwiftCommandHandler ` is when reading the definition of `commandsAndHandlers`.

ahoppen commented on Sources/SourceKitLSP/Swift/SwiftLanguageService.swift:
```diff
@@ -797,15 +802,27 @@ extension SwiftLanguageService {
       }
     }
+
+    if (canInlineMacro) {
```
```suggestion
    if canInlineMacro {
```

ahoppen commented on Sources/SourceKitLSP/Swift/SwiftLanguageService.swift:
```diff
@@ -797,15 +802,27 @@ extension SwiftLanguageService {
+    var refactorActions = cursorInfoResponse.refactorActions.compactMap {
       do {
         let lspCommand = try $0.asCommand()
+        canInlineMacro = $0.title == "Inline Macro"
```
Instead of checking the macro title, what do you think about comparing the `actionString`. That should be more stable than the title.

ahoppen commented on Sources/SourceKitLSP/Swift/SwiftLanguageService.swift:
```diff
@@ -797,15 +802,27 @@ extension SwiftLanguageService {
     }
+
+    if (canInlineMacro) {
+      let expandMacroCommand = try! ExpandMacroCommand(positionRange: params.range, textDocument: params.textDocument)
```
We shouldn’t use `try!` because we don’t want to crash the entire SourceKit-LSP server just because some request failed. Checking really quickly, I think we can just make `asCommand` not throw. If that’s not possible, we should use `orLog` to log the error and then just not add the code action.

ahoppen commented on Sources/SourceKitLSP/Swift/SwiftLanguageService.swift:
```diff
@@ -922,29 +939,19 @@ extension SwiftLanguageService {
-        reason = " reason: \(failureReason)"
-      } else {
-        reason = ""
+
```
Superfluous newline.

ahoppen commented on Sources/SourceKitLSP/Swift/SwiftLanguageService.swift:
```diff
@@ -922,29 +939,19 @@ extension SwiftLanguageService {
     }
-    return edit.encodeToLSPAny()
+
+    throw ResponseError.unknown("semantic refactoring: unknown command \(req.command)")
```
`ExecuteCommand` no longer is just about semantic refactorings, right? I’d just drop the `semantic refactoring: ` prefix from the log message.

ahoppen commented on Sources/SourceKitLSP/Swift/MacroExpansionEdit.swift:
```diff
@@ -0,0 +1,57 @@
+import SourceKitD
+
+/// Represents a macro expansion as an edit. Notionally, a subclass of `TextEdit`
+public struct MacroExpansionEdit: ResponseType, Hashable, Sendable {
```
Does this type have to be `public`? Also, I don’t think it has to conform to `ResponseType` because we never return it as a response to an LSP request – we always encode it to `LSPAny` first.

lokesh-tr replied:
Yeah right

lokesh-tr replied:
Wait, I realised that the tests are using the `MacroExpansionEdit`. So, it has to be `public`? or should we not mark this `public` and try some workarounds in the tests?

ahoppen replied:
Ah, I missed that. Let’s mark it as `@_spi(Testing) public` then and `@_spi(Testing) import SourceKitLSP` in the test.

ahoppen commented on Sources/SourceKitLSP/Swift/SemanticRefactoring.swift:
```diff
@@ -103,41 +58,44 @@ extension SemanticRefactoringError: CustomStringConvertible {
-      keys.actionUID: self.sourcekitd.api.uid_get_from_cstr(refactorCommand.actionString)!,
-      keys.compilerArgs: await self.buildSettings(for: snapshot.uri)?.compilerArgs as [SKDRequestValue]?,
-    ])
+    guard let refactorCommand = refactorCommand as? SemanticRefactorCommand else {
+      throw ResponseError.unknown("refactorCommand is not a SemanticRefactorCommand")
+    }
```
Similar to my comment on the macro expansion, should we make this function take a `SemanticRefactorCommand` instead of `any SwiftCommand`?

ahoppen commented on Sources/SourceKitLSP/Swift/SwiftLanguageService.swift:
```diff
@@ -922,29 +939,19 @@ extension SwiftLanguageService {
-      } else {
-        reason = ""
+
+    let commandsAndHandlers: [(command: any SwiftCommand.Type, handler: SwiftCommandHandler)] = [
+      (command: SemanticRefactorCommand.self, handler: semanticRefactoring),
+      (command: ExpandMacroCommand.self, handler: expandMacro),
+    ]
```
Instead of looping here, do you think it would be simpler to just have a quick `if` series? That way we could change `semantRefactoring` and `expandMacro` to take concrete types instead of `any SwiftCommand`

Something like

```swift
if let command = req.swiftCommand(ofType: SemanticRefactorCommand.self) {
  return try await semanticRefactoring(command)
} else if let command = req.swiftCommand(ofType: ExpandMacroCommand.self) {
  return try await expandMacro(command)
}
```

ahoppen commented on Sources/SourceKitLSP/Swift/MacroExpansion.swift:
```diff
@@ -0,0 +1,94 @@
+///
+/// Wraps the information returned by sourcekitd's `semantic_refactoring` request, such as the necessary macro expansion edits.
+struct MacroExpansion: Refactoring {
+
```
Superfluous newline

ahoppen commented on Sources/SourceKitLSP/Swift/MacroExpansion.swift:
```diff
@@ -0,0 +1,94 @@
+      )
+    }
+  }
+
```
Superfluous newline

ahoppen commented on Sources/SourceKitLSP/Swift/MacroExpansion.swift:
```diff
@@ -0,0 +1,94 @@
+      MacroExpansionEdit(
+        range: refactoringEdit.startPosition..<refactoringEdit.endPosition,
+        newText: refactoringEdit.newText,
+        bufferName: refactoringEdit.bufferName ?? ""  // TODO: handle bufferName if nil
```
I would just log a fault and then ignore that edit.

ahoppen commented on Sources/SourceKitLSP/Swift/MacroExpansion.swift:
```diff
@@ -0,0 +1,94 @@
+        throw ResponseError.unknown("Unable to write macro expansion to file path: \"\(macroExpansionFilePath.path)\"")
+      }
+
+      let req = ShowDocumentRequest(uri: macroExpansionDocURI, selection: macroEdit.range)
+      let response = try await sourceKitLSPServer.sendRequestToClient(req)
+      if !response.success {
+        logger.error("client refused to show document for \(expansion.title, privacy: .public)!")
+      }
```
I would prefer to send this request to the client asynchronously (ie. wrap it in a `Task`) because otherwise it’s an easy way to deadlock SourceKit-LSP in the future. I don’t think it can happen right now but depending on how `MessageHandlingDependencyTracker` gets changed in the future it’s not completely unreasonable to assume that as part of the `ShowDocumentRequest` to the client, the client will send a request to SourceKit-LSP that cannot be executed until the `ExecuteCommandRequest` finishes executing. But that can’t finish executing until the `ShowDocumentRequest` has finished, so deadlock.

lokesh-tr replied:
Yeah that makes sense

ahoppen commented on Tests/SourceKitLSPTests/ExecuteCommandTests.swift:
```diff
@@ -136,6 +136,95 @@ final class ExecuteCommandTests: XCTestCase {
+        }
+
+        @main
+        struct MyMacroPlugin: CompilerPlugin {
+            let providingMacros: [Macro.Type] = [
+                StringifyMacro.self,
+            ]
+        }
```
Nitpick: I would indent this using 2 spaces for consistency.

ahoppen commented on Tests/SourceKitLSPTests/ExecuteCommandTests.swift:
```diff
@@ -136,6 +136,95 @@ final class ExecuteCommandTests: XCTestCase {
+        }
+        """,
+      ],
+      manifest: SwiftPMTestProject.macroPackageManifest
```
You can remove the other test case that uses `SwiftPMTestProject.macroPackageManifest` now. That test was basically a no-op that was just there to make sure that `SwiftPMTestProject.macroPackageManifest` works correctly. And we have this test to do that now.

ahoppen commented on Tests/SourceKitLSPTests/ExecuteCommandTests.swift:
```diff
@@ -136,6 +136,95 @@ final class ExecuteCommandTests: XCTestCase {
+
+    let (uri, positions) = try project.openDocument("MyMacroClient.swift")
+
+    let args = ExpandMacroCommand(
+      positionRange: positions["1️⃣"]..<positions["2️⃣"],
+      textDocument: TextDocumentIdentifier(uri)
+    )
```
I think it would make sense to also check that we show macro expansions at the zero character ranges (aka positions) at `1️⃣#stringify` and `#1️⃣stringify`.

ahoppen commented on Tests/SourceKitLSPTests/ExecuteCommandTests.swift:
```diff
@@ -136,6 +136,95 @@ final class ExecuteCommandTests: XCTestCase {
+      return
+    }
+
+    XCTAssertEqual(resultArray.count, 1)
+    XCTAssertEqual(resultArray[0].newText, "(1 + 2, \"1 + 2\")")
```
We have `.only` for exactly this pattern 😉 This also prevents the test from crashing if `resultArray` is empty.

```suggestion
    XCTAssertEqual(resultArray.only?.newText, "(1 + 2, \"1 + 2\")")
```

lokesh-tr replied:
Cool

ahoppen commented on Tests/SourceKitLSPTests/ExecuteCommandTests.swift:
```diff
@@ -136,6 +136,95 @@ final class ExecuteCommandTests: XCTestCase {
+    XCTAssertEqual(resultArray.count, 1)
+    XCTAssertEqual(resultArray[0].newText, "(1 + 2, \"1 + 2\")")
+
+    let generatedMacroExpansionsPath = SourceKitLSPServer.Options.testDefault.generatedMacroExpansionsPath
+
+    XCTAssert(
+      (try? generatedMacroExpansionsPath.appending(component: resultArray[0].bufferName).asURL
+        .checkResourceIsReachable()) ?? false
+    )
+
```
Instead of inspecting the internals of the macro expansion here, what do you think about the following (which would also be closer to what editors do):
- In the handler for the `ShowDocumentRequest` request, we record the URL of the generated buffer
- We use an XCTest expectation to make sure we have received the `ShowDocumentRequest` here.
- And then we inspect the contents of that file.

ahoppen commented on Tests/SourceKitLSPTests/ExecuteCommandTests.swift:
```diff
@@ -136,6 +136,95 @@ final class ExecuteCommandTests: XCTestCase {
+      )
+    )
+
+    XCTAssertTrue(fileContents.contains("(1 + 2, \"1 + 2\")"))
```
Nit
```suggestion
    XCTAssert(fileContents.contains("(1 + 2, \"1 + 2\")"))
```

ahoppen reviewed (commented):
Very impressive. This is looking great. All my comments are minor / code style related. The overall design looks great!

yaroslavyaroslav commented:
Sorry to disturbing you folks in such a busy time, but just to clarify for myself does this implementation would add `ExpandSnippet` code action option in along with `ExtractMethod` on a client side for free or there's be some work on their side required?

lokesh-tr commented:
> Sorry to disturbing you folks in such a busy time, but just to clarify for myself does this implementation would add `ExpandSnippet` code action option in along with `ExtractMethod` on a client side for free or there's be some work on their side required?

@yaroslavyaroslav well I don't think this implementation would add anything related to what you have asked. There's this `ExecuteCommandTests.testRangeSemanticRefactoring` which tests whether extract method works (already implemented, not specific to this PR). But am not sure if the same is implemented on the client side and am not sure sourcekitd sends an `ExpandSnippet` code action to the LSP.

Perhaps @ahoppen can answer?

@yaroslavyaroslav I guess you can open an issue and discuss this with him to not spam this PR since its not related to this PR.

ahoppen commented:
Sorry, I don’t fully understand your question @yaroslavyaroslav. If we want to continue discussing it, could you please file an issue so we can have a dedicated thread for it?

lokesh-tr commented:
Based on the meeting which we had yesterday:

- [ ] Address Review Comments

- [ ] Wrap it in experimental feature

- [ ] Rename generated macro file name to have line numbers

- [ ] Implement custom peek requests

lokesh-tr commented:
I was little skeptical with my implementation of `Refactoring` (thought I would replace it with a better name at the end but forgot) and the passing around `T.Type`. I did think about using associated typesbut that didn't fit well with my array-looping based implementation of `executeCommandRequest`. Now that I changed it to if-elseif, the associate type implementation is trivial.

I have resolved all your comments. Kindly review and let me know.

ahoppen commented on Tests/SourceKitLSPTests/ExecuteCommandTests.swift:
```diff
@@ -136,6 +136,104 @@ final class ExecuteCommandTests: XCTestCase {
+        (try? url.checkResourceIsReachable()) ?? false
+      )
+
+      let fileContents = try XCTUnwrap(try? String(contentsOf: url, encoding: .utf8))
```
I think you can save the `XCTUnwrap` dance here and just throw if it fails. The benefit would be that this also shows the error why the string couldn’t load the file contents when it fails.

```suggestion
      let fileContents = try String(contentsOf: url, encoding: .utf8)
```

ahoppen commented on Tests/SourceKitLSPTests/ExecuteCommandTests.swift:
```diff
@@ -136,6 +136,104 @@ final class ExecuteCommandTests: XCTestCase {
+        positionRange: positions["1️⃣"]..<positions["3️⃣"],
+        textDocument: TextDocumentIdentifier(uri)
+      ),
+      ExpandMacroCommand(
+        positionRange: positions["2️⃣"]..<positions["3️⃣"],
+        textDocument: TextDocumentIdentifier(uri)
+      ),
```
Could you also add tests for `1️⃣..<1️⃣` and `2️⃣..<2️⃣`? That should cover cases when you are invoking the command from a cursor without selecting text.

ahoppen reviewed (commented):
Looking great. I have a few minor comments still but everything is very local.

lokesh-tr commented on Sources/SourceKitLSP/Swift/MacroExpansion.swift:
```diff
@@ -0,0 +1,98 @@
+        let req = ShowDocumentRequest(uri: macroExpansionDocURI, selection: macroEdit.range)
+        let response = try await sourceKitLSPServer.sendRequestToClient(req)
+        if !response.success {
+          logger.error("client refused to show document for \(expansion.title, privacy: .public)!")
```
I don't think that the `ShowDocumentResponse` as per the LSP spec has a property for the reason behind the error. Am I missing anything? 🤔

ahoppen replied:
Oh, sorry, I was mistaking this with the client returning an error response. And looking at it again, I notice that if we get an error from `sendRequestToClient`, we just swallow it. We should log that by wrapping `sendRequestToClient` in `orLog`.

lokesh-tr commented:
@ahoppen I have addressed the review comments. When possible, let me know if there is anything else which I should fix. 😌 

I have also made a new commit.
Instead of creating generated files with buffer name in `GeneratedMacroExpansions` directory, this will now create a new subdirectory for each buffer within the `GeneratedMacroExpansions`. Within that subdirectory, will exist a file with the name as original source file containing the macro expansion.

This is inline with what we discussed in the meeting of adding a GitHub style name notation. The only difference is that instead of having just the starting line number in the file name like `SourceFile.swift#L3`, this implementation adds position range to the file name as `SourceFile.swift#L3C2-L4C5`. This is based on my finding that GitHub supports a url fragment which denotes the starting and ending position separated with an hyphen, each denoted by the letters L and C representing Line number and Column Number respectively.

The start and end position are equal when new lines are added while expanding. Similarly, they are unequal when new lines are replacing old lines while expanding. And thus, allows the user to just look at the name of the expanded file name to decide whether it is adding or replacing new lines.

lokesh-tr commented:
Based on the discussion about my findings which we had,
I changed the file naming scheme as follows: `SourceFileName_LaCb-LcCd.swift`

The benefit of the above naming is that:
1. The user can be clear about whether the macro is adding new lines or replacing with new lines just by looking at the position range.
2. No use of `#` means, no need to worry about fragments, escaping the sign, etc.
3. Instead of putting the position range indicator after the file extension, we put it before the extension, so that the code editors can understand the actual file extension.

ahoppen commented on Sources/SourceKitLSP/Swift/MacroExpansion.swift:
```diff
@@ -0,0 +1,122 @@
+    self.uri = uri
+    self.edits = refactoringEdits.compactMap { refactoringEdit in
+      if refactoringEdit.bufferName.isEmpty {
+        logger.error("Unable to retrieve some parts of the expansion")
```
This gave me an excuse to write some document about when to use which logging level. I think `fault` would be more applicable here because if we don’t have a buffer name that’s a bug in sourcekitd.

https://github.com/apple/sourcekit-lsp/blob/main/Documentation/Logging.md

ahoppen commented on Sources/SourceKitLSP/Swift/MacroExpansion.swift:
```diff
@@ -0,0 +1,122 @@
+  /// and then makes `ShowDocumentRequest` to the client side for each
+  /// expansion to be displayed.
+  ///
+  /// - Parameters:
+  ///   - expandMacroCommand: The `ExpandMacroCommand` that triggered this
+  /// request.
+  ///
+  /// - Returns:
+  ///   - an `[MacroExpansionEdit]` with the necessary edits and buffer name as
+  /// a `LSPAny`
```
Some formatting suggestions that should make this read a little better, I think.

```suggestion
  /// - Parameters:
  ///   - expandMacroCommand: The `ExpandMacroCommand` that triggered this request.
  ///
  /// - Returns: An `[MacroExpansionEdit]` with the necessary edits and buffer name as a `LSPAny`
```

ahoppen commented on Sources/SourceKitLSP/Swift/MacroExpansion.swift:
```diff
@@ -0,0 +1,122 @@
+    }
+
+    guard let sourceFileURL = expandMacroCommand.textDocument.uri.fileURL else {
+      throw ResponseError.unknown("Given URL is not a file URL")
```
Nit: LSP communicates in URIs instead of URLs

```suggestion
      throw ResponseError.unknown("Given URI is not a file URL")
```

ahoppen commented on Sources/SourceKitLSP/Swift/RefactoringEdit.swift:
```diff
@@ -0,0 +1,57 @@
+import LanguageServerProtocol
+import SourceKitD
+
+/// Represents an edit from semantic refactor response. Notionally, a subclass /// of `TextEdit`
```
```suggestion
/// Represents an edit from semantic refactor response. Notionally, a subclass of `TextEdit`
```

ahoppen commented on Sources/SourceKitLSP/Swift/MacroExpansion.swift:
```diff
@@ -0,0 +1,122 @@
+    let expansion = try await self.refactoring(expandMacroCommand)
+
+    for macroEdit in expansion.edits {
+      let macroExpansionBufferDirectoryName = macroEdit.bufferName.dropLast(6)  // buffer name without ".swift"
```
Should we check that the buffer name has a `.swift` suffix? Otherwise I can imagine that we could get very odd failure scenarios if the buffer name does not end in `.swift`.

ahoppen commented on Sources/SourceKitLSP/Swift/MacroExpansion.swift:
```diff
@@ -0,0 +1,122 @@
+    for macroEdit in expansion.edits {
+      let macroExpansionBufferDirectoryName = macroEdit.bufferName.dropLast(6)  // buffer name without ".swift"
+
+      let macroExpansionBufferDirectoryURL = self.generatedMacroExpansionsPath.appendingPathComponent(
+        String(macroExpansionBufferDirectoryName)
+      )
```
I generally prefer to put `appendingPathComponent` on a new line because it scales if you need to add more components to URLs.

```suggestion
      let macroExpansionBufferDirectoryURL = self.generatedMacroExpansionsPath
        .appendingPathComponent(String(macroExpansionBufferDirectoryName))
```

ahoppen commented on Sources/SourceKitLSP/Swift/MacroExpansion.swift:
```diff
@@ -0,0 +1,122 @@
+      let macroExpansionDocURI = DocumentURI(macroExpansionFilePath)
+
+      do {
+        try macroEdit.newText.write(to: macroExpansionFilePath, atomically: true, encoding: String.Encoding.utf8)
```
You can drop the `String.Encoding`.

```suggestion
        try macroEdit.newText.write(to: macroExpansionFilePath, atomically: true, encoding: .utf8)
```

ahoppen commented on Sources/SourceKitLSP/Swift/MacroExpansion.swift:
```diff
@@ -0,0 +1,122 @@
+      }
+
+      Task {
+        let req = ShowDocumentRequest(uri: macroExpansionDocURI, selection: macroEdit.range)
```
I think this is the only place that uses `macroExpansionDocURI`, so you could just change it to the following and remove `macroExpansionDocURI`.

```suggestion
        let req = ShowDocumentRequest(uri: DocumentURI(macroExpansionFilePath), selection: macroEdit.range)
```

ahoppen commented on Sources/SourceKitLSP/Swift/Refactoring.swift:
```diff
@@ -0,0 +1,130 @@
+  ///   - keys: The sourcekitd key set to use for looking up into `dict`.
+  init?(_ title: String, _ dict: SKDResponseDictionary, _ snapshot: DocumentSnapshot, _ keys: sourcekitd_api_keys) {
+    guard let categorizedEdits: SKDResponseArray = dict[keys.categorizedEdits] else {
+      logger.error("categorizedEdits doesn't exist in response dictionary")
```
Similar to my comment above, I think this could be a `fault`.

```suggestion
      logger.fault("categorizedEdits doesn't exist in response dictionary")
```

Same for the other `error`s you are introducing in this file.

ahoppen commented on Sources/SourceKitLSP/Swift/Refactoring.swift:
```diff
@@ -0,0 +1,130 @@
+        return true
+      }
+      edits.forEach { _, edit in
+        // The LSP is zero based, but semantic_refactoring is one based.
```
I think this comment makes more sense before the lines where you do the `- 1` calculations.

ahoppen commented on Sources/SourceKitLSP/Swift/Refactoring.swift:
```diff
@@ -0,0 +1,130 @@
+    }
+
+    guard !refactoringEdits.isEmpty else {
+      logger.error("No refactors found")
```
```suggestion
      logger.error("No refactoring edits found")
```

ahoppen commented on Sources/SourceKitLSP/Swift/Refactoring.swift:
```diff
@@ -0,0 +1,130 @@
+  ///
+  /// - Parameters:
+  ///   - refactorCommand: The semantic `RefactorCommand` that triggered this request.
+  /// - Returns:
+  ///   - The `RefactoringResponse` as `T`
```
Also, I’m not entirely sure what you mean by _The `RefactoringResponse` as `T`_. I would just go for something like: _The response of the refactoring_.

---

Nit: You don’t need a sub-bullet point for the description of the return value.

```suggestion
  /// - Returns: The `RefactoringResponse` as `T`
```

ahoppen commented on Sources/SourceKitLSP/Swift/RefactoringEdit.swift:
```diff
@@ -0,0 +1,57 @@
+  /// The new text.
+  public var newText: String
+
+  public var bufferName: String
```
Should we add a doc comment for `bufferName` because I think it’s not entirely obvious what it is? Something like 

> If the new text of the edit should not be applied to the original source file but to a separate buffer, a fake name for that buffer. For example for expansion of macros, this is `@` followed by the mangled name of the macro expansion, followed by `.swift`.

ahoppen commented on Sources/SourceKitLSP/Swift/SemanticRefactoring.swift:
```diff
@@ -103,41 +60,42 @@ extension SemanticRefactoringError: CustomStringConvertible {
-  ///   - refactorCommand: The semantic refactor `Command` that triggered this request.
+  ///   - semanticRefactorCommand: The `SemanticRefactorCommand` that triggered this request.
+  ///
+  /// - Returns:
+  ///   - a `WorkspaceEdit` with the necessary refactors as a `LSPAny`
```
```suggestion
  /// - Returns: A `WorkspaceEdit` with the necessary refactors as a `LSPAny`
```

ahoppen commented on Tests/SourceKitLSPTests/ExecuteCommandTests.swift:
```diff
@@ -136,6 +137,123 @@ final class ExecuteCommandTests: XCTestCase {
+
+      let request = ExecuteCommandRequest(command: command.command, arguments: command.arguments)
+
+      let expectation = XCTestExpectation(description: "Handle Show Document Request")
```
`self.expectation` also raises in error if the expectation is never awaited, which is quite useful.

```suggestion
      let expectation = self.expectation(description: "Handle Show Document Request")
```

ahoppen commented on Tests/SourceKitLSPTests/ExecuteCommandTests.swift:
```diff
@@ -136,6 +137,123 @@ final class ExecuteCommandTests: XCTestCase {
+        "Wrong macro expansion. Failed for position range between \(positionMarker.start) and \(positionMarker.end)"
+      )
+
+      await fulfillment(of: [expectation], timeout: 10.0)
```
`fulfillmentOfOrThrow` is a useful utility that uses `defaultTimeout` and throws an error if it times out, which is useful because it’s very likely that all the following assertions will fail if the expectation times out.

```suggestion
      try await fulfillmentOfOrThrow([expectation])
```

ahoppen reviewed (commented):
I have a few more minor comments inline.

lokesh-tr commented:
The last commit which I made tries to convert the sequential edit from sourcekitd into `ConcurrentEdits` as we discussed, but it fails the `testFreestandingMacroExpansion` test, the reason being the line numbers aren't matching after i convert them to concurrent edits.

I would like to have you help me debug this issue.

lokesh-tr commented:
I removed the `ConcurrentEdits` commit from here... I will wrap the current state of freestanding macro expansion as such in  experimental features and will make a seperate PR for concurrent edits and attached macro expansion

lokesh-tr commented:
@ahoppen I wrapped the feature under `ExperimentalFeatures`. Let me have your comments so that I can fix them by tomorrow and also squash the commits. Hope we will merge this soon.

Further, I will follow up with a new PR for the `ConcurrentEdits`.

ahoppen commented on Sources/SourceKitLSP/Swift/MacroExpansion.swift:
```diff
@@ -0,0 +1,118 @@
+      }
+    }
+
+    return expansion.edits.encodeToLSPAny()
```
Do we even need to return any edits here? I would expect that we have already handled the showing of the macro expansion buffers and the client shouldn’t need to deal with the edits returned from the `ExecuteCommandRequest`, right? Also, since it’s `LSPAny`, I would expect that the client doesn’t even know how to interpret it. Or am I missing something?

Edit: I now see that the tests rely on `ExecuteCommand` returning the edits. If we don’t need `expandMacro` and `semanticRefactoring` to return any results for editors, I think we should change this in a follow-up PR to get this one merged.

lokesh-tr replied:
The way the semantic refactoring for inline macro command was implemented is such that the testing part utilises the result. Since it was already accepted into the codebase, I thought it was a convention and implemented macro expansion command in a similar manner.

I guess it would be better to merge this PR and make a fix for this is in a follow-up PR as you said.

ahoppen replied:
Filed https://github.com/apple/sourcekit-lsp/issues/1498 so we don’t forget about it.

ahoppen commented on Sources/SourceKitLSP/Swift/Refactoring.swift:
```diff
@@ -0,0 +1,129 @@
+          RefactoringEdit(
+            range: startPosition..<endPosition,
+            newText: textWithSnippets,
+            bufferName: edit[keys.bufferName] ?? ""
```
Do you think a cleaner design would be to make `RefactoringEdit.bufferName` an optional? Then we wouldn’t need to nil-coalesce here and I think `nil` represents the absence of a buffer name better than an empty string.

lokesh-tr replied:
Sure, and I would also highlight the fact that it is ok for bufferName to be nil (i.e. not passed as a response from sourcekitd) if the refactoring edit is meant to represent the removal of the macro annotation from the original source file (here, newText will be empty). Its only a fault of sourcekitd if the refactoring edit's buffer name is nil when the refactoring edit represents an expansion or replacement and has an associated newText value which won't be empty.

ahoppen commented on Sources/SourceKitLSP/Swift/SemanticRefactoring.swift:
```diff
@@ -103,41 +60,41 @@ extension SemanticRefactoringError: CustomStringConvertible {
+  /// Handles the `SemanticRefactorCommand`.
   ///
-  /// Wraps the information returned by sourcekitd's `semantic_refactoring` request, such as the necessary edits and placeholder locations.
+  /// Makes request to sourcekitd and wraps the result into a
```
```suggestion
  /// Sends a request to sourcekitd and wraps the result into a
```

ahoppen commented on Sources/SourceKitLSP/Swift/SemanticRefactoring.swift:
```diff
@@ -103,41 +60,41 @@ extension SemanticRefactoringError: CustomStringConvertible {
-  ///   - refactorCommand: The semantic refactor `Command` that triggered this request.
+  ///   - semanticRefactorCommand: The `SemanticRefactorCommand` that triggered this request.
+  ///
+  /// - Returns: A `WorkspaceEdit` with the necessary refactors as a `LSPAny`
```
I think this comment is out-of-date, right? This function doesn’t return a `WorkspaceEdit`. It actually got me wondering: Do we need to return anything here at all? Similar to my comment on `macroExpansion`, I wouldn’t expect the client to interpret the `LSPAny` and we already handled the edits by sending `ApplyEditRequest` to the client.

lokesh-tr replied:
Here, I would like to highlight the reply which I made to your comment on `macroExpansion`

ahoppen replied:
Just to double-check: You mean [this comment](https://github.com/apple/sourcekit-lsp/pull/1436#discussion_r1647371565), right? Saying that we should change this to not return the edits, at which point the comment will go away, right?

ahoppen commented on Sources/SourceKitLSP/Swift/SwiftLanguageService.swift:
```diff
@@ -815,15 +831,31 @@ extension SwiftLanguageService {
-        return nil
+    var canInlineMacro = false
+
+    let showMacroExpansionsIsEnabled = await self.sourceKitLSPServer?.options.experimentalFeatures.contains(
+      .showMacroExpansions
+    )
```
Formatting nitpick again 😉 Also, I would just nil-coalesce here so we don’t need to deal with `showMacroExpansionsIsEnabled == nil` below

```suggestion
    let showMacroExpansionsIsEnabled = await self.sourceKitLSPServer?.options.experimentalFeatures
      .contains(.showMacroExpansions) ?? false
```

ahoppen commented on Sources/SourceKitLSP/Swift/SwiftLanguageService.swift:
```diff
@@ -815,15 +831,31 @@ extension SwiftLanguageService {
+
+    var refactorActions = cursorInfoResponse.refactorActions.compactMap {
+      let lspCommand = $0.asCommand()
+      canInlineMacro = $0.actionString == "source.refactoring.kind.inline.macro"
+
+      if !(showMacroExpansionsIsEnabled != nil && showMacroExpansionsIsEnabled!) {
+        canInlineMacro = false
       }
```
Do you think this would read cleaner as follows? I found the setting of `canInlineMacro` to `true` just to revert it back to `false` a little confusing. I also noticed that we would set `canInlineMacro` back to `false` if `source.refactoring.kind.inline.macro` was not the last refactoring action.

```suggestion
      if !canInlineMacro, showMacroExpansionsIsEnabled {
        canInlineMacro = $0.actionString == "source.refactoring.kind.inline.macro"
      }
```

ahoppen commented on Tests/SourceKitLSPTests/ExecuteCommandTests.swift:
```diff
@@ -136,6 +137,124 @@ final class ExecuteCommandTests: XCTestCase {
+        """,
+      ],
+      manifest: SwiftPMTestProject.macroPackageManifest,
+      serverOptions: .init(swiftPublishDiagnosticsDebounceDuration: 0, experimentalFeatures: [.showMacroExpansions])
```
I would prefer to build on top of `.testDefault` so that if we change `.testDefault`, the changes get applied here as well. The current pattern is that at the top of the test function we do something like

```swift
var serverOptions = SourceKitLSPServer.Options.testDefault
serverOptions.experimentalFeatures.insert(.showMacroExpansions)
```

ahoppen reviewed (commented):
I only have very minor comments now. Could you address them and then we can get the PR merged. My comment about making `expandMacro` and `semanticRefactoring` to not return the edits is more wide-reaching and I think we should look into that in a follow-up PR to get this one merged.

ahoppen reviewed (approved):
Looking great! Let’s get this 🚢ed.

ahoppen commented:
@swift-ci Please test

ahoppen commented:
@swift-ci Please test Windows

ahoppen commented on Sources/SourceKitLSP/Swift/ExpandMacroCommand.swift:
On Windows, we build SourceKit-LSP using CMake instead of SwiftPM. For this, these new files need to be added to [CMakeLists.txt](https://github.com/apple/sourcekit-lsp/blob/main/Sources/SourceKitLSP/CMakeLists.txt)

ahoppen commented:
@swift-ci Please test

ahoppen commented:
@swift-ci Please test Windows
